feat: [PKD-004][PKD-007] [UI][Implementation] Pokemon Information Pag…

Enrickzz opened the pull request:
…e: Evolution

- added evolution_handler
- added evolution_chain_model
- added pokemon_evolution_tab (with vm and connector)
- added GetEvolution action on home_page_actions
- added evolution in app_state

https://user-images.githubusercontent.com/61441278/182092828-bd8b7e3e-1b88-41cd-bc9d-693b09b16f74.mp4

Enrickzz committed:
feat: [PKD-004][PKD-007] [UI][Implementation] Pokemon Information Page: Evolution

- added evolution_handler
- added evolution_chain_model
- added pokemon_evolution_tab (with vm and connector)
- added GetEvolution action on home_page_actions
- added evolution in app_state

raelsj commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,114 @@
+            Column(
+              children: thisPokeEvo.evolutionTwo
+                  .map(
+                    (e) => Container(
```
DRY

dbenedictos commented on lib/api/handlers/evolution_handler.dart:
```diff
@@ -0,0 +1,59 @@
+import 'package:http/http.dart' as http;
+import 'package:pokedex_async_redux/api/models/species_model.dart';
+
+class EvoluttionHandler {
```
typo

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,114 @@
+                ],
+              ),
+            ),
+            Container(
```
please also handle cases where the pokemon has no evolution
![Screenshot 2022-08-01 at 3 28 50 PM](https://user-images.githubusercontent.com/47196695/182096273-133eb1ca-e5b3-4622-8c7e-2bb1b323a08e.png)
.

Enrickzz committed:
PR Feedback

-Resolve typo (EvolutionHandler)
-added reusable widget for pokemon_evolution_tab
-UI handling on no evolution pokemons

dbenedictos commented on lib/api/models/evolution_chain_model.dart:
```diff
@@ -0,0 +1,16 @@
+  factory EvolutionChain({
+    @JsonKey(name: 'evolutionOne') Species? evolutionOne,
+    @Default(<Species>[]) @JsonKey(name: 'evolutionTwo') List<Species> evolutionTwo,
+    @JsonKey(name: 'evolutionThree') Species? evolutionThree,
```
There are pokemons with more than one third evolution. i.e. 
- 
![Screenshot 2022-08-01 at 3 35 33 PM](https://user-images.githubusercontent.com/47196695/182101792-62a6f788-1539-44d3-a58a-dcb8fc447ca8.png)

-
![Screenshot 2022-08-01 at 4 01 08 PM](https://user-images.githubusercontent.com/47196695/182101822-353f5286-fafc-47eb-9b74-1cc58bdb1353.png)

Enrickzz committed:
PR Feedback

- evo3 to List

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,46 @@
+                ? Container(
+                    width: 50,
+                    height: 50,
+                    child: Icon(Icons.arrow_right_outlined),
```
the icon widget has a size property, can you use that instead of wrapping it with a container?

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/widgets/pokemon_evolution_tab_widget.dart:
```diff
@@ -0,0 +1,35 @@
+    return Container(
+      height: 80,
+      width: 80,
+      child: Stack(
```
why stack? For your UI I think this is not applicable.

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evultion_tab_vm.dart:
```diff
@@ -0,0 +1,32 @@
+  @override
+  Vm fromStore() {
+    return PokemonEvolutionVm(
+      evolutionState: _getPageState(),
```
trailing comma

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evultion_tab_vm.dart:
```diff
@@ -0,0 +1,32 @@
+
+class PokemonEvolutionVm extends Vm {
+  PokemonEvolutionVm({
+    required this.evolutionState,
```
trailing comma

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,46 @@
+              width: 80,
+              child: PokemonEvolutionWidget(thisSpecies: thisPokeEvo.evolutionOne),
+            ),
+            thisPokeEvo.evolutionTwo.isNotEmpty
```
`if(condition)`
`   Widget()`

doesn't make sense to return a sizedBox. Either add the widget if it meets the condition or don't add it.

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,46 @@
+                    child: Icon(Icons.arrow_right_outlined),
+                  )
+                : SizedBox(),
+            Column(
```
still Dry imo

Enrickzz committed:
PR Feedback

-evolution_tab clean up (DRY code issue)
-traling commas

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/widgets/pokemon_evolution_tab_widget.dart:
```diff
@@ -0,0 +1,56 @@
+import 'package:pokedex_async_redux/utilities/constants.dart';
+import 'package:pokedex_async_redux/utilities/extensions.dart';
+
+class PokemonEvolutionWidget extends StatelessWidget {
```
match your file name with your widget name

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,34 @@
+    return Padding(
+      padding: EdgeInsets.all(5),
+      child: SingleChildScrollView(
+        child: Wrap(
```
centre the children

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evultion_tab_vm.dart:
```diff
@@ -0,0 +1,28 @@
+
+class PokemonEvolutionVmFactory extends VmFactory<AppState, PokemonEvolutionConnector> {
+  @override
+  Vm fromStore() {
```
=>

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/widgets/pokemon_evolution_tab_widget.dart:
```diff
@@ -0,0 +1,56 @@
+    return Column(
+      children: thisSpecies
+          .map(
+            (thisSpecies) => Container(
```
suggestion: rename this to `thisSpecie` so that it implies a single item from your list.

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/widgets/pokemon_evolution_tab_widget.dart:
```diff
@@ -0,0 +1,56 @@
+                  if (arrowCheck)
+                    Positioned(
+                      left: -55,
+                      child: Container(
```
remove the container. Assign the bottom and top of the positioned widget to 0, so it centres vertically

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/widgets/pokemon_evolution_tab_widget.dart:
```diff
@@ -0,0 +1,56 @@
+                    width: 50,
+                    height: 50,
+                  ),
+                  Positioned(
```
try adjusting your simulator's font size, The texts overlap each other. And the arrows aren't centred anymore. Better to just have a simple column here than to use stack. For the arrows just put it in your evolution tab. so it's always at the centre of two pokemons
![Screenshot 2022-08-01 at 5 32 02 PM](https://user-images.githubusercontent.com/47196695/182119196-a1de5150-563a-4934-be69-424924b2fd8a.png)

Enrickzz committed:
PR Feedback

-Stack to column
-Centered evolution chaiin
-Minor UI fixes

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,39 @@
+      padding: EdgeInsets.all(5),
+      child: SingleChildScrollView(
+        child: Center(
+          child: Wrap(
```
play with `alignment` & `crossAxisAlignment` so that you can remove the Center widget

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,39 @@
+            children: [
+              PokemonEvolutionTabWidget(thisSpecies: <Species>[thisPokeEvo.evolutionOne!]),
+              if (thisPokeEvo.evolutionTwo.isNotEmpty)
+                Container(
```
I assume you're wrapping the Icon with a Container for the spacing. Play around the `alignment` property of the wrap widget. I bet you can find something that will apply even spaces around its children

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,39 @@
+                  width: 50,
+                  child: Icon(Icons.arrow_right_outlined),
+                ),
+              PokemonEvolutionTabWidget(thisSpecies: thisPokeEvo.evolutionTwo),
```
if the list passed to this widget is empty, it will result in an empty column. Just like my comment earlier, either add the widget or don't add it. 😛

Enrickzz committed:
PR Feedback

-remove unecessary widgets
-use properties of wrap widget

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,31 @@
+          alignment: WrapAlignment.center,
+          children: [
+            PokemonEvolutionTabWidget(thisSpecies: <Species>[thisPokeEvo.evolutionOne!]),
+            if (thisPokeEvo.evolutionTwo.isNotEmpty) Icon(Icons.arrow_right_outlined),
```
`if(thisPokeEvo.evolutionTwo.isNotEmpty) ... [Icon(Icons.arrow_right_outlined), PokemonEvolutionTabWidget(thisSpecies: thisPokeEvo.evolutionTwo),],`

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/evolution_tab/pokemon_evolution_tab.dart:
```diff
@@ -0,0 +1,31 @@
+            PokemonEvolutionTabWidget(thisSpecies: <Species>[thisPokeEvo.evolutionOne!]),
+            if (thisPokeEvo.evolutionTwo.isNotEmpty) Icon(Icons.arrow_right_outlined),
+            if (thisPokeEvo.evolutionTwo.isNotEmpty) PokemonEvolutionTabWidget(thisSpecies: thisPokeEvo.evolutionTwo),
+            if (thisPokeEvo.evolutionThree.isNotEmpty) Icon(Icons.arrow_right_outlined),
```
same here. the `...[]` is called the spread operator

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/pokemon_tab_view.dart:
```diff
@@ -30,27 +31,39 @@ class TabView extends StatelessWidget {
-                    child: Text(
-                      pokemonDetailsTabs[0],
-                      style: TextStyle(color: Colors.black),
+                    child: FittedBox(
```
DRY

Enrickzz committed:
PR Feedback

-if spread operator ( ...[] )
-separated a widget for tabs (Pokemon info tabs)

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/pokemon_info_tabs.dart:
```diff
@@ -0,0 +1,20 @@
+  const PokemonInfoTabs({
+    required this.tab,
+  });
+  final String tab;
```
naming sucks

dbenedictos replied:
jk

dbenedictos replied:
make this descriptive. What part of the tab?, title? name?

Enrickzz committed:
PR Feedback

-descriptive var naming
